### DISCUSSION TITLE
Does csrfguard support multiple web servers in load balancing environment?

### DISCUSSION THREAD

**Discussion by sumaharjan** [OWASP/www-project-csrfguard#146] [Q&A] [answered]
csrfguards saves token in memory under ConcurrentHashMap, thus, I doubt if it works in multi stack environment. The request can be served in any of web server which might not have token in memory , hence might treat even legitimate request as threat. Please let me know if there is anything different to my thought. Thanks in advance

**Comment by forgedhallpass** [comment, marked as answer]
> 1. If you need session replication you could try tying the tokens back to the session, so they would be automatically replicated (although it would require changes in the library and you would need to find a way for the [`TokenHolder`](https://github.com/OWASP/www-project-csrfguard/blob/master/csrfguard/src/main/java/org/owasp/csrfguard/token/storage/TokenHolder.java) interface not to depend on HTTP sessions directly, because that would not make sense in case of stateless web apps).
>
> 2. Have you considered/tried moving the tokens to a database or to an in-memory cache instead of relying on the [`InMemoryTokenHolder`](https://github.com/OWASP/www-project-csrfguard/blob/c787303d4613247f0dc2b8df875d867e80a95ca7/csrfguard/src/main/java/org/owasp/csrfguard/token/storage/impl/InMemoryTokenHolder.java)?

Taken from https://github.com/OWASP/www-project-csrfguard/discussions/121#discussioncomment-3445751
